HelloWorld_ST_Sensors/user/gps/gps.cpp

```diff
@@ -0,0 +1,152 @@
+
+        if (strcmp(temp_token, "$GPRMC") == 0)
+        {
+            gps_nmea_type = GPS_GPRMC;
+        }
+
+        // Not processing GPGGA since it does not have an Active/Valid field to determine the validity of NMEA string
+        // else if (strcmp(temp_token, "$GPGGA") == 0)
+        // {
+        //     gps_nmea_type = GPS_GPGGA;
+        // }
+
+        if (gps_nmea_type != GPS_INVALID_NMEA)
+        {
+            while (temp_token != NULL)
+            {
+                temp_token = strtok(NULL, ",");
+                nmea_string_token_index++;
+                switch (gps_nmea_type)
+                {
+                case GPS_GPGGA:
+                    switch (nmea_string_token_index)
+                    {
+                    case GPGGA_LATITUDE_INDEX:
+                        strcpy(gps_coordinates_data.latitude, temp_token);
+                        break;
+                    case GPGGA_LATITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.lat_dir, temp_token);
+                        break;
+                    case GPGGA_LONGITUDE_INDEX:
+                        strcpy(gps_coordinates_data.longitude, temp_token);
+                        break;
+                    case GPGGA_LONGITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.long_dir, temp_token);
+                        break;
+                    }
+                    break;
+                case GPS_GPRMC:
+                    switch (nmea_string_token_index)
+                    {
+                    case GPRMC_VALIDITY_INDEX:
+                        strcpy(gps_coordinates_data.validity, temp_token);
+                        // Consider NMEA strings that are GPRMC and Active Navigation Receiver Warning
+                        if (strcmp(temp_token, "A") != 0)
+                        {
+                            gps_nmea_type = GPS_INVALID_NMEA;
+                        }
+                        break;
+                    case GPRMC_LATITUDE_INDEX:
+                        strcpy(gps_coordinates_data.latitude, temp_token);
+                        break;
+                    case GPRMC_LATITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.lat_dir, temp_token);
+                        break;
+                    case GPRMC_LONGITUDE_INDEX:
+                        strcpy(gps_coordinates_data.longitude, temp_token);
+                        break;
+                    case GPRMC_LONGITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.long_dir, temp_token);
+                        break;
+                    }
+                    break;
+                }
+            }
+        }
+        return gps_nmea_type;
+    }
+
+    uint8_t gps_c::extract_gpgga_or_gprmc_line(char *gps_buff)
+    {
+        uint8_t gps_nmea_type = GPS_INVALID_NMEA;
+        char valid_gps_string[100] = {0};
+        uint16_t counter = 0;
+        for (uint16_t i = 0; i < UART_BUF_LEN; i++)
+        {
```

Comment:
As mentioned in the interrupt callback, you are iterating over the entire UART_BUF_LEN
But when you receive the first $ value you are parsing that and exiting
There can be multiple $ tokens received we should not need to miss any